src/main/java/seedu/address/model/assignment/AssignmentList.java

```diff
@@ -162,102 +163,228 @@ public void calculateScheduleIntensity(int numDays) {
+            }
+
+            LocalDateTime midnightToday = LocalDateTime.of(currDateTime.toLocalDate().plusDays(1), LocalTime.MIDNIGHT);
+
+            // Number of hours that have already been allocated to query day before this assignment
+            float prevHoursAlrAllocatedToToday = cumulativeHoursPerDayList.get(0) - hoursPerDayList.get(0);
+            // Amount of time on query day that can be allocated to this current assignment
+            float hoursLeftToToday = Math.abs(ChronoUnit.HOURS.between(currDateTime, midnightToday))
+                - prevHoursAlrAllocatedToToday;
+
+            LocalDateTime midnightDeadline = LocalDateTime.of(deadline.toLocalDate(), LocalTime.MIDNIGHT);
+
+            // Number of hours that have already been allocated to the day of the deadline before this assignment
+            float prevHoursAlrAllocatedToDeadline = cumulativeHoursPerDayList.get(noOfDaysBetween)
+                - hoursPerDayList.get(noOfDaysBetween);
+            // Amount of time on the day the assignment is due that can be allocated to this current assignment
+            float hoursLeftToDeadlineDay = Math.max(Math.abs(ChronoUnit.HOURS.between(midnightDeadline, deadline))
+                - prevHoursAlrAllocatedToDeadline, 0);
+
+            // Excess hours allocated to query date will be redistributed across the next few days up to deadline
+            // Excess hours allocated to deadline day will be redistributed over previous days, starting from query date
+            if (hoursLeftToToday < hoursPerDayList.get(0)
+                && hoursLeftToDeadlineDay < hoursPerDayList.get(noOfDaysBetween)) {
+
+                float excessToday = hoursPerDayList.get(0) - hoursLeftToToday;
+                float excessDeadline = hoursPerDayList.get(noOfDaysBetween) - hoursLeftToDeadlineDay;
 
-            for (int k = 0; k < Math.min(hoursPerDayList.size(), noOfDaysBetween + 1); k++) {
-                hoursPerDayList.set(k, hoursPerDayList.get(k) + hoursToAdd);
+                cumulativeHoursPerDayList.set(0, cumulativeHoursPerDayList.get(0) - excessToday);
+                hoursPerDayList.set(0, hoursPerDayList.get(0) - excessToday);
+
+                cumulativeHoursPerDayList.set(noOfDaysBetween, cumulativeHoursPerDayList.get(noOfDaysBetween)
+                    - excessDeadline);
+                hoursPerDayList.set(noOfDaysBetween, hoursPerDayList.get(noOfDaysBetween) - excessDeadline);
+
+                hoursAdded = (excessToday + excessDeadline) / (noOfDaysBetween - 1);
+                redistributeHours(1, noOfDaysBetween, hoursAdded, cumulativeHoursPerDayList, hoursPerDayList);
+
+            } else if (hoursLeftToToday < hoursPerDayList.get(0)) {
+                float excessToday = hoursPerDayList.get(0) - hoursLeftToToday;
+
+                cumulativeHoursPerDayList.set(0, cumulativeHoursPerDayList.get(0) - excessToday);
+                hoursPerDayList.set(0, hoursPerDayList.get(0) - excessToday);
```

Comment:
do i have to extract this out into another function arh since it repeats itself quite a few times O: